plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTypeManager.java

```diff
@@ -380,6 +402,69 @@ private Optional<ColumnMapping> convertToTrinoType(Field field)
         }
     }
 
+    public List<ColumnMetadata> convertToTrinoType(List<String> names, List<String> types)
+    {
+        checkArgument(names.size() == types.size(), "Mismatched column names and types");
+
+        ImmutableList.Builder<ColumnMetadata> columns = ImmutableList.builder();
+        for (int i = 0; i < names.size(); i++) {
+            String type = types.get(i);
+            TypeInfo typeInfo = getTypeInfoFromTypeString(type);
+            TypeSignature typeSignature = toTypeSignature(typeInfo);
+            columns.add(new ColumnMetadata(names.get(i), typeManager.getType(typeSignature)));
+        }
+        return columns.build();
+    }
+
+    private TypeSignature toTypeSignature(TypeInfo typeInfo)
+    {
+        switch (typeInfo.getCategory()) {
+            case PRIMITIVE:
+                Type primitiveType = fromPrimitiveType((PrimitiveTypeInfo) typeInfo);
+                if (primitiveType == null) {
+                    break;
+                }
+                return primitiveType.getTypeSignature();
+            case ARRAY:
+                ArrayTypeInfo arrayTypeInfo = (ArrayTypeInfo) typeInfo;
+                TypeSignature elementType = toTypeSignature(arrayTypeInfo.getListElementTypeInfo());
+                return arrayType(typeParameter(elementType));
+            case STRUCT:
+                StructTypeInfo structTypeInfo = (StructTypeInfo) typeInfo;
+                List<TypeInfo> fieldTypes = structTypeInfo.getAllStructFieldTypeInfos();
+                List<String> fieldNames = structTypeInfo.getAllStructFieldNames();
+                verify(fieldTypes.size() == fieldNames.size(), "Invalid BigQuery struct type: %s", typeInfo);
+                return rowType(Streams.zip(
+                                fieldNames.stream(),
+                                fieldTypes.stream().map(this::toTypeSignature),
+                                TypeSignatureParameter::namedField)
+                        .collect(Collectors.toList()));
+        }
+        throw new TrinoException(NOT_SUPPORTED, format("Unsupported BigQuery type: %s", typeInfo));
+    }
+
+    @Nullable
+    private Type fromPrimitiveType(PrimitiveTypeInfo typeInfo)
+    {
+        return switch (typeInfo.getStandardSqlTypeName()) {
+            case BOOL -> BOOLEAN;
+            case BYTES -> VARBINARY;
+            case INT64 -> INTEGER;
+            case FLOAT64 -> REAL;
+            case DATE -> DATE;
+            case DATETIME -> TIMESTAMP_MICROS;
+            case NUMERIC, BIGNUMERIC -> {
+                DecimalTypeInfo decimalTypeInfo = (DecimalTypeInfo) typeInfo;
+                yield createDecimalType(decimalTypeInfo.precision(), decimalTypeInfo.scale());
+            }
+            case STRING -> VARCHAR;
+            case GEOGRAPHY -> VARCHAR;
+            case JSON -> jsonType;
+            case TIMESTAMP -> BIGINT;
+            default -> null;
```

Comment:
throw new TrinoException("Unsupported type"); then return null handling at the call-site